src/keystore/providers/SnapProvider.ts

```diff
@@ -0,0 +1,88 @@
+import { KeystoreProviderUnavailableError } from './errors'
+import { Keystore } from './../interfaces'
+import { KeystoreProvider, KeystoreProviderOptions } from './interfaces'
+import { defaultSnapOrigin, SnapKeystore } from '../SnapKeystore'
+
+export type Snap = {
+  permissionName: string
+  id: string
+  version: string
+  initialPermissions: Record<string, unknown>
+}
+
+export type GetSnapsResponse = Record<string, Snap>
+
+const isFlask = async () => {
+  const provider = window.ethereum
+
+  try {
+    const clientVersion = await provider?.request({
+      method: 'web3_clientVersion',
+    })
+
+    const isFlaskDetected = (clientVersion as string[])?.includes('flask')
+
+    return Boolean(provider && isFlaskDetected)
+  } catch {
+    return false
+  }
+}
+
+const getSnaps = async (): Promise<GetSnapsResponse> => {
+  return (await window.ethereum.request({
+    method: 'wallet_getSnaps',
+  })) as unknown as GetSnapsResponse
+}
+
+const getSnap = async (version?: string): Promise<Snap | undefined> => {
+  try {
+    const snaps = await getSnaps()
+
+    return Object.values(snaps).find(
+      (snap) =>
+        snap.id === defaultSnapOrigin && (!version || snap.version === version)
+    )
+  } catch (e) {
+    console.warn('Failed to obtain installed snap', e)
+    return undefined
+  }
+}
+
+const initSnap = async () => {
+  await window.ethereum.request({
```

Comment:
I have a special `init` method on the Snap, which is the only time it asks for wallet signatures (and only if the keys are not available in MM storage)